### PR TITLE
feat(dynamic-field): dynamic fields are added to detail pages.

### PR DESCRIPTION
**Pull request by sanbornsen**
### [ ] What does this PR do? _[Mandatory]_

* Added dynamic fields in the detail and quick-preview page.
* Added restrictions on updating dynamic fields and regular static fields for non-collaborators.
* Dropdown field issue is fixed.
* Only `enum` and  `markup` fields will be rendered other than static fields. 

### [ ] What issue/task does this PR references? _[Mandatory]_

* https://openshift.io/openshiftio/openshiftio/plan/detail/1950
* https://github.com/openshiftio/openshift.io/issues/2736

#### [ ] Are the Release Notes included?

No

#### [ ] Is/Are the @mention(s) to reviewer(s) included

**Note:** This PR is compaitable with current backend but best experience will be visible only once the update dynamic field error is resolved. @kwk is working on that for the backend side. 

**Note:** Markup fields won't get rendered until it has the proper value as per contract.

### CODE REVIEW

**Comment by Raunak1203** [comment]
[test]

**Comment by Raunak1203** [comment]
@sanbornsen unit tests are failing. Could you please check

**Comment by sanbornsen** [comment]
[test]

**Comment by Raunak1203** [comment]
@sanbornsen functional tests are failing. Could you update selector for Iteration and Area dropdown ?

**Comment by Raunak1203** [comment]
@sanbornsen 
1) area, iteration dropdown header shows "default Header text" . It Should be "Add area", "Add Iteration?
2) If iteration name is long then should it be truncated? In prod it is truncated
Please see the screenshot
screenshot from 2018-05-11 15-14-17

**Comment by sanbornsen** [comment]
[test]

**Comment by sanbornsen** [comment]
[test]

**Comment by Raunak1203** [comment]
[test]

**Comment by sanbornsen** [comment]
[test]

File: src/app/components_ngrx/dynamic-field/dynamic-field.component.ts [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+      this.dropdownSelectedItems =
+        this.dropdownMenuItems.filter(v => v.key === this.fieldValue.value);
+    }
+
+    // if it's an boolean type
+    // set the dropdown values
+    // to support the dropdown
+    if (this.fieldValue.field.type.kind === 'boolean') {
+      this.dropdownMenuItems = this.extractBooleanKeyValues()
```
Comment (vikram-raj) [inline src/app/components_ngrx/dynamic-field/dynamic-field.component.ts]: Semicolon missing.

File: src/app/components_ngrx/work-item-detail/work-item-detail.component.html [hunk 2]
```diff
@@ -221,92 +220,119 @@
...
+                <!-- <typeahead-dropdown #areaSelectbox
                   *ngIf="!loadingArea"
                   id="area-dropdown"
-                  noValueLabel="None" 
+                  noValueLabel="None"
                   [values]="areas"
                   (onFocus)="focusArea()"
                   (onUpdate)="areaUpdated($event)">
-                </typeahead-dropdown>
+                </typeahead-dropdown> -->
```
Comment (vikram-raj) [inline src/app/components_ngrx/work-item-detail/work-item-detail.component.html]: Please remove this commented component. We will not use this anymore.

File: src/app/components_ngrx/work-item-detail/work-item-detail.component.html [hunk 3]
```diff
@@ -221,92 +220,119 @@
...
-
           <!-- Iteration -->
           <div class="form-group">
             <section class="f8-detail__area">
               <label class="pull-left padding-left-15 control-label">
                 {{workItem.type.fields['system.iteration']?.label || 'Label'}}
               </label>
               <div class="f8-detail__content">
-                <span *ngIf="loadingIteration"
+                  <span *ngIf="loadingIteration"
```
Comment (vikram-raj) [inline src/app/components_ngrx/work-item-detail/work-item-detail.component.html]: Fix indentation.

File: src/app/components_ngrx/work-item-detail/work-item-detail.component.html [hunk 4]
```diff
@@ -221,92 +220,119 @@
...
-                    #iterationSelectbox 
-                    noValueLabel="None" 
-                    [values]="iterations"
-                    (onFocus)="focusIteration()" 
-                    (onUpdate)="iterationUpdated($event)">
-                  </typeahead-dropdown>
+                    class="iteration-dropdown"
+                    #iterationSelectbox noValueLabel="None" [values]="iterations"
+                    (onFocus)="focusIteration()" (onUpdate)="iterationUpdated($event)">
+                  </typeahead-dropdown> -->
```
Comment (vikram-raj) [inline src/app/components_ngrx/work-item-detail/work-item-detail.component.html]: Remove this commented component.

File: src/app/reducers/detail-work-item.reducer.ts [hunk 5]
```diff
@@ -19,6 +20,10 @@ export const DetailWorkItemReducer: ActionReducer<DetailWorkItemState> =
         return {...action.payload};
       }
 
+      case WorkItemActions.UPDATE_ERROR: {
+        return {...state};
```
Comment (vikram-raj) [inline src/app/reducers/detail-work-item.reducer.ts]: Why are you updating the state on error?

**Comment by vikram-raj** [comment]
screenshot from 2018-05-22 11-54-34

Getting default icon for all state of the work item in scrum template.

**Comment by sanbornsen** [comment]
> Can we use common-selector component in place of asignee-selector and label-selector component?

@vikram-raj no we can't it has different complexity.

File: src/app/components/work-item-detail/work-item-detail.component.ts [hunk 6]
```diff
@@ -291,7 +291,7 @@ export class WorkItemDetailComponent implements OnInit, OnDestroy {
 
           // init dynamic form
           if (this.workItem.relationships.baseType.data.attributes) {
-            this.dynamicFormGroup = this.workItemTypeControlService.toFormGroup(this.workItem);
+            // this.dynamicFormGroup = this.workItemTypeControlService.toFormGroup(this.workItem);
```
[the remark this answers is not in the document]
Comment (sanbornsen) [reply src/app/components/work-item-detail/work-item-detail.component.ts]: Do not consider any changes inside `components` directory.

File: src/app/reducers/detail-work-item.reducer.ts (hunk 5 above)
Comment (sanbornsen) [reply src/app/reducers/detail-work-item.reducer.ts]: On work item update error we need to reset the field and there is no way for the component to know if it failed in effects. So we need this for this scenario.

**Comment by sanbornsen** [comment]
> Resolution and severity fields are not intended properly.

@vikram-raj They are totally intended. This is the PR all about, to begin with. DYNAMIC FIELDS.

**Comment by sanbornsen** [comment]
> Getting default icon for all state of the work item in scrum template.

Not relevant for the PR. This is an update needs to be done in `ngx-widgets` level.

**Comment by vikram-raj** [comment]
[test]

**Pull request merged**